File: vane-core/src/main/java/org/oddlama/vane/core/lang/LangManager.java
package org.oddlama.vane.core.lang;

import static org.reflections.ReflectionUtils.getAllFields;

import java.io.File;
import java.lang.annotation.Annotation;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.function.Function;
import java.util.logging.Level;
import java.util.stream.Collectors;
import org.bukkit.configuration.file.YamlConfiguration;
import org.oddlama.vane.annotation.lang.LangMessage;
import org.oddlama.vane.annotation.lang.LangMessageArray;
import org.oddlama.vane.annotation.lang.LangVersion;
import org.oddlama.vane.core.YamlLoadException;
import org.oddlama.vane.core.module.Module;
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator;

public class LangManager {

    Module<?> module;
    private List<LangField<?>> langFields = new ArrayList<>();
    LangVersionField fieldVersion;

    public LangManager(Module<?> module) {
        this.module = module;
        compile(module, s -> s);
    }

    public long expectedVersion() {
        return module.annotation.langVersion();
    }

    private boolean hasLangAnnotation(Field field) {
        for (var a : field.getAnnotations()) {
            if (a.annotationType().getName().startsWith("org.oddlama.vane.annotation.lang.Lang")) {
                return true;
            }
        }
        return false;
    }

    private void assertFieldPrefix(Field field) {
        if (!field.getName().startsWith("lang")) {
            throw new RuntimeException("Language fields must be prefixed lang. This is a bug.");
        }
    }

    private LangField<?> compileField(Object owner, Field field, Function<String, String> mapName) {
        assertFieldPrefix(field);

        // Get the annotation
        Annotation annotation = null;
        for (var a : field.getAnnotations()) {
            if (a.annotationType().getName().startsWith("org.oddlama.vane.annotation.lang.Lang")) {
                if (annotation == null) {
                    annotation = a;
                } else {
                    throw new RuntimeException("Language fields must have exactly one @Lang annotation.");
                }
            }
        }
        assert annotation != null;
        final var atype = annotation.annotationType();

        // Return a correct wrapper object
        if (atype.equals(LangMessage.class)) {
            return new LangMessageField(module, owner, field, mapName, (LangMessage) annotation);
        } else if (atype.equals(LangMessageArray.class)) {
            return new LangMessageArrayField(module, owner, field, mapName, (LangMessageArray) annotation);
        } else if (atype.equals(LangVersion.class)) {
            if (owner != module) {
                throw new RuntimeException("@LangVersion can only be used inside the main module. This is a bug.");
            }
            if (fieldVersion != null) {
                throw new RuntimeException(
                    "There must be exactly one @LangVersion field! (found multiple). This is a bug."
                );
            }
            return fieldVersion = new LangVersionField(module, owner, field, mapName, (LangVersion) annotation);
        } else {
            throw new RuntimeException("Missing LangField handler for @" + atype.getName() + ". This is a bug.");
        }
    }

    private boolean verifyVersion(File file, long version) {
        if (version != expectedVersion()) {
            module.log.severe(file.getName() + ": expected version " + expectedVersion() + ", but got " + version);

            if (version == 0) {
                module.log.severe("Something went wrong while generating or loading the configuration.");
                module.log.severe("If you are sure your configuration is correct and this isn't a file");
                module.log.severe(
                    "system permission issue, please report this to https://github.com/oddlama/vane/issues"
                );
            } else if (version < expectedVersion()) {
                module.log.severe("This language file is for an older version of " + module.getName() + ".");
                module.log.severe("Please update your file or use an officially supported language file.");
            } else {
                module.log.severe("This language file is for a future version of " + module.getName() + ".");
                module.log.severe("Please use the correct file for this version, or use an officially");
                module.log.severe("supported language file.");
            }

            return false;
        }

        return true;
    }

    @SuppressWarnings("unchecked")
    public void compile(Object owner, Function<String, String> mapName) {
        // Compile all annotated fields
        langFields.addAll(
            getAllFields(owner.getClass())
                .stream()
                .filter(this::hasLangAnnotation)
                .map(f -> compileField(owner, f, mapName))
                .toList()
        );

        if (owner == module && fieldVersion == null) {
            throw new RuntimeException("There must be exactly one @LangVersion field! (found none). This is a bug.");
        }
    }

    @SuppressWarnings("unchecked")
    public <T> T getField(String name) {
        var field = langFields
            .stream()
            .filter(f -> f.getName().equals(name))
            .findFirst()
            .orElseThrow(() -> new RuntimeException("Missing lang field lang" + name));

        try {
            return (T) field;
        } catch (ClassCastException e) {
            throw new RuntimeException("Invalid lang field type for lang" + name, e);
        }
    }

    public boolean reload(File file) {
        // Load file
        var yaml = YamlConfiguration.loadConfiguration(file);

        // Check version
        final var version = yaml.getLong("Version", -1);
        if (!verifyVersion(file, version)) {
            return false;
        }

        try {
            // Check languration for errors
            for (var f : langFields) {
                f.checkLoadable(yaml);
            }

            for (var f : langFields) {
                f.load(module.namespace(), yaml);
            }
        } catch (YamlLoadException e) {
            module.log.log(Level.SEVERE, "error while loading '" + file.getAbsolutePath() + "'", e);
            return false;
        }
        return true;
    }

    public void generateResourcePack(final ResourcePackGenerator pack, YamlConfiguration yaml, File langFile) {
        var langCode = yaml.getString("ResourcePackLangCode");
        if (langCode == null) {
            throw new RuntimeException("Missing yaml key: ResourcePackLangCode");
        }
        var errors = new LinkedList<YamlLoadException.Lang>();
        for (var f : langFields) {
            try {
                f.addTranslations(pack, yaml, langCode);
            } catch (YamlLoadException.Lang e) {
                errors.add(e);
            } catch (YamlLoadException e) {
                module.log.log(Level.SEVERE, "Unexpected YAMLLoadException: ", e);
            }
        }
        if (errors.size() > 0) {
            final String erroredLangNodes = errors
                .stream()
                .map(Throwable::getMessage)
                .collect(Collectors.joining("\n\t\t"));
            module.log.log(
                Level.SEVERE,
                "The following errors were identified while adding translations from \n\t" +
                langFile.getAbsolutePath() +
                " \n\t\t" +
                erroredLangNodes
            );
        }
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/core/lang/LangManager.kt
package org.oddlama.vane.core.lang

import org.bukkit.configuration.file.YamlConfiguration
import org.oddlama.vane.annotation.lang.LangMessage
import org.oddlama.vane.annotation.lang.LangMessageArray
import org.oddlama.vane.annotation.lang.LangVersion
import org.oddlama.vane.core.YamlLoadException
import org.oddlama.vane.core.module.Module
import org.oddlama.vane.core.resourcepack.ResourcePackGenerator
import org.reflections.ReflectionUtils
import java.io.File
import java.lang.reflect.Field
import java.util.logging.Level

/**
 * Discovers, validates, and loads language fields for a module.
 *
 * @param module the module whose language entries are managed.
 */
class LangManager(var module: Module<*>) {
    /**
     * All compiled language field handlers.
     */
    private val langFields: MutableList<LangField<*>> = mutableListOf()

    /**
     * The compiled `@LangVersion` field, if discovered.
     */
    var fieldVersion: LangVersionField? = null

    init {
        compile(module) { it ?: "" }
    }

    /**
     * Returns the expected language file version from module metadata.
     */
    fun expectedVersion(): Long = module.annotation.langVersion

    /**
     * Returns whether a field has any supported language annotation.
     */
    private fun hasLangAnnotation(field: Field): Boolean =
        field.annotations.any { it.annotationClass.java.name.startsWith("org.oddlama.vane.annotation.lang.Lang") }

    /**
     * Ensures language fields follow the required `lang` prefix convention.
     */
    private fun assertFieldPrefix(field: Field) {
        if (!field.name.startsWith("lang"))
            throw RuntimeException("Language fields must be prefixed lang. This is a bug.")
    }

    /**
     * Compiles a reflected language field into a concrete [LangField] handler.
     */
    private fun compileField(owner: Any?, field: Field, mapName: (String?) -> String): LangField<*> {
        assertFieldPrefix(field)

        val annotation = field.annotations
            .filter { it.annotationClass.java.name.startsWith("org.oddlama.vane.annotation.lang.Lang") }
            .also { require(it.size <= 1) { "Language fields must have exactly one @Lang annotation." } }
            .firstOrNull() ?: error("No @Lang annotation found on field ${field.name}")

        return when (val atype = annotation.annotationClass.java) {
            LangMessage::class.java      -> LangMessageField(module, owner, field, mapName, annotation as LangMessage)
            LangMessageArray::class.java -> LangMessageArrayField(module, owner, field, mapName, annotation as LangMessageArray)
            LangVersion::class.java      -> {
                require(owner === module) { "@LangVersion can only be used inside the main module. This is a bug." }
                require(fieldVersion == null) { "There must be exactly one @LangVersion field! (found multiple). This is a bug." }
                LangVersionField(module, owner, field, mapName, annotation as LangVersion).also { fieldVersion = it }
            }
            else -> throw RuntimeException("Missing LangField handler for @${atype.name}. This is a bug.")
        }
    }

    /**
     * Verifies a language file version against the expected module version.
     */
    private fun verifyVersion(file: File, version: Long): Boolean {
        if (version == expectedVersion()) return true

        module.log.severe("${file.name}: expected version ${expectedVersion()}, but got $version")
        when {
            version == 0L -> {
                module.log.severe("Something went wrong while generating or loading the configuration.")
                module.log.severe("If you are sure your configuration is correct and this isn't a file")
                module.log.severe("system permission issue, please report this to https://github.com/oddlama/vane/issues")
            }
            version < expectedVersion() -> {
                module.log.severe("This language file is for an older version of ${module.name}.")
                module.log.severe("Please update your file or use an officially supported language file.")
            }
            else -> {
                module.log.severe("This language file is for a future version of ${module.name}.")
                module.log.severe("Please use the correct file for this version, or use an officially")
                module.log.severe("supported language file.")
            }
        }
        return false
    }

    /**
     * Discovers and compiles language fields from an owner instance.
     *
     * @param owner the object containing language fields.
     * @param mapName maps Java field names to YAML paths.
     */
    fun compile(owner: Any, mapName: (String?) -> String) {
        langFields.addAll(
            ReflectionUtils.getAllFields(owner.javaClass)
                .filter { hasLangAnnotation(it) }
                .map { compileField(owner, it, mapName) }
        )
        if (owner === module && fieldVersion == null)
            throw RuntimeException("There must be exactly one @LangVersion field! (found none). This is a bug.")
    }

    /**
     * Returns a compiled language field by mapped name.
     */
    fun getField(name: String?): LangField<*> =
        langFields.firstOrNull { it.name == name }
            ?: throw RuntimeException("Missing lang field lang$name")

    /**
     * Reloads language values from a language file.
     *
     * @param file the language YAML file.
     * @return `true` on success.
     */
    fun reload(file: File): Boolean {
        val yaml = YamlConfiguration.loadConfiguration(file)
        val version = yaml.getLong("Version", -1)
        if (!verifyVersion(file, version)) return false

        try {
            langFields.forEach { it.checkLoadable(yaml) }
            langFields.forEach { it.load(module.namespace(), yaml) }
        } catch (e: YamlLoadException) {
            module.log.log(Level.SEVERE, "error while loading '${file.absolutePath}'", e)
            return false
        }
        return true
    }

    /**
     * Adds translatable entries from a language file to a resource pack generator.
     *
     * @param pack the target resource pack generator.
     * @param yaml the loaded language configuration.
     * @param langFile the source language file.
     */
    fun generateResourcePack(pack: ResourcePackGenerator?, yaml: YamlConfiguration, langFile: File) {
        val langCode = yaml.getString("ResourcePackLangCode")
            ?: throw RuntimeException("Missing yaml key: ResourcePackLangCode")

        val errors = mutableListOf<YamlLoadException.Lang>()
        for (f in langFields) {
            try {
                f.addTranslations(pack, yaml, langCode)
            } catch (e: YamlLoadException.Lang) {
                errors.add(e)
            } catch (e: YamlLoadException) {
                module.log.log(Level.SEVERE, "Unexpected YAMLLoadException: ", e)
            }
        }
        if (errors.isNotEmpty()) {
            val erroredLangNodes = errors.joinToString("\n\t\t") { it.message }
            module.log.log(
                Level.SEVERE,
                "The following errors were identified while adding translations from \n\t${langFile.absolutePath} \n\t\t$erroredLangNodes"
            )
        }
    }
}
